langserver/handlers/did_open.go

```diff
@@ -15,5 +19,51 @@ func TextDocumentDidOpen(ctx context.Context, params lsp.DidOpenTextDocumentPara
 	}
 
 	f := ilsp.FileFromDocumentItem(params.TextDocument)
-	return fs.Open(f)
+	err = fs.Open(f)
+	if err != nil {
+		return err
+	}
+
+	cf, err := lsctx.RootModuleCandidateFinder(ctx)
+	if err != nil {
+		return err
+	}
+
+	rootDir, _ := lsctx.RootDirectory(ctx)
+
+	candidates := cf.RootModuleCandidatesByPath(f.Dir())
+	if len(candidates) == 0 {
+		msg := fmt.Sprintf("No root module found for %s"+
+			" functionality may be limited", f.Filename())
+		return jrpc2.ServerPush(ctx, "window/showMessage", lsp.ShowMessageParams{
+			Type:    lsp.MTWarning,
+			Message: msg,
+		})
+	}
+	if len(candidates) > 1 {
+		// TODO: Suggest specifying explicit root modules?
+
+		msg := fmt.Sprintf("Alternative root modules found for %s (%s), picked: %s",
+			f.Filename(), renderCandidates(rootDir, candidates[1:]),
+			renderCandidate(rootDir, candidates[0]))
+		return jrpc2.ServerPush(ctx, "window/showMessage", lsp.ShowMessageParams{
+			Type:    lsp.MTWarning,
+			Message: msg,
+		})
+	}
+
+	return nil
+}
+
+func renderCandidates(rootDir string, candidatePaths []string) string {
+	for i, p := range candidatePaths {
+		// This helps displaying shorter, but still relevant paths
+		candidatePaths[i] = renderCandidate(rootDir, p)
+	}
+	return strings.Join(candidatePaths, ", ")
+}
+
+func renderCandidate(rootDir, path string) string {
+	return strings.TrimPrefix(
```

Comment:
I wonder if there is a better `filepath` func to do this to get the relative path? String manipulation feels wrong but maybe its fine?

Reply:
Yeah, it does not feel quite right, but I couldn't find `filepath.TrimPrefix` or anything similar.